File: Array_Operation.rb
def element_at(arr, index)

  if(index < 0 || index > arr.size)
    return "Invalid index";
  end
  puts "Element at index #{index} is:"
  puts "#{arr[index]}"

  return arr[index]
end

def inclusive_range(arr, start_pos, end_pos)

  if(start_pos < 0 || start_pos > arr.size || end_pos < 0 || end_pos > arr.size)
    return "Invalid arguments"
  end
  return arr[start_pos...end_pos]

end

def non_inclusive_range(arr, start_pos, end_pos)

  if(start_pos < 0 || start_pos > arr.size || end_pos < 0 || end_pos > arr.size)
    return "Invalid arguments"
  end
  return arr[start_pos...end_pos-1]

end

def start_and_length(arr, start_pos, length)

  if(start_pos < 0 || start_pos > arr.size || length < 0 || length > arr.size)
    return "Invalid arguments"
  end

  if((start_pos+length) < arr.length)
     arr[start_pos, length]
  else
    return "Length Out of Bound"
  end
end

arr = [9, 5, 1, 2, 3, 4, 0, -1]



puts element_at(arr, 0)

puts "\ninclusive range Start Position#{0} and End Position#{4}"
puts inclusive_range(arr, 0, 4)


puts "\nnon_inclusive range"
puts non_inclusive_range(arr, 1, 5)

puts "\nstart_and_length"
puts start_and_length(arr, 5, 4)

File: Python/Q2/ArrayOperation.py
class MyException(Exception):
    pass
class ArrayOperations:
    def __init__(self, list) -> None:
        self.list = list
        pass

    def element_at(self, index):
        if(type(index) == int):
            if(index < 0 or index > len(self.list)):
                raise MyException("Index not in range")
            else:
                return list[index]
        else:
            raise MyException("Only integers are allowed")

    def inclusive_range(self, start_pos, end_pos):
        if(start_pos < 0 or start_pos > len(self.list) or end_pos < 0 or end_pos > len(self.list) or end_pos < start_pos):
            raise MyException("Index not in range")
        else:
            for i in range(start_pos, end_pos):
                return list[start_pos:end_pos+1]

    def non_inclusive_range(self, start_pos, end_pos):
        if(start_pos < 0 or start_pos > len(self.list) or end_pos < 0 or end_pos > len(self.list) or end_pos < start_pos):
            raise MyException("Index not in range")
        else:
            return list[start_pos:end_pos]

    def start_and_length(self, start_pos, length):
        if(start_pos < 0 or start_pos > len(self.list) or length < 0 or length > len(self.list)):
            raise MyException("Index not in range")
        elif((start_pos+length) < len(self.list)):
            return list[start_pos:length+1]
        else:
            raise MyException("Length Out of Bound")


list = [9, 5, 1, 2, 3, 4, 0, -1]
index = 1
array_op = ArrayOperations(list)
print(array_op.element_at(index))
print(array_op.inclusive_range(1,3))
print(array_op.non_inclusive_range(1,3))
print(array_op.start_and_length(1,3))
